compiler/src/dotty/tools/dotc/transform/TupleRewrites.scala

```diff
@@ -0,0 +1,215 @@
+package dotty.tools.dotc
+package transform
+
+import ast.Trees._
+import core.Constants.Constant
+import core.Contexts.Context
+import core.Definitions.MaxImplementedTupleArity
+import core.StdNames._
+import core.Symbols._
+import core.Types._
+import transform.TreeTransforms.{MiniPhaseTransform, TransformerInfo}
+
+/** Local rewrites for tuple expressions. Nested apply and unapply trees coming
+ *  from desugaring into single apply/unapply nodes on DottyTupleN/LargeTuple.
+ */
+class TupleRewrites extends MiniPhaseTransform {
+  import ast.tpd._
+  import TupleRewrites._
+
+  def phaseName: String = "tupleRewrites"
+
+  override def checkPostCondition(tree: Tree)(implicit ctx: Context): Unit =
+    tree match {
+      case Select(ident, _) if ident.symbol == defn.TupleConsModule && start != end =>
+        assert(false, s"Reference to TupleCons comming from desugaring survived tupleRewrites.")
+      case _ => ()
+    }
+
+  /** Rewrites `TupleCons(a, TupleCons(b, ..., TNit))` to implementation specific constructors.
+   *
+   *  Below `MaxImplementedTupleArity`, they become `DottyTuple$i(a, b, ...)`.
+   *  Above `MaxImplementedTupleArity`, they become `LargeTuple(Array.apply(a, b, ...)`.
+   *
+   *  Note that because of bottom up traversal, the transformation of a tuple constructor of size `N`
+   *  will go thought this transformation `N` times, thus generating `N` `TupleCons(a, opt)` where `opt`
+   *  is the optimized transformation for previous arity.
+   */
+  override def transformApply(tree: Apply)(implicit ctx: Context, info: TransformerInfo): Tree = {
+    // Matches a tree with shape `TupleCons.apply(head, tail)` where `tail` itself a tuple
+    // with statically known lenght (Unit, DottyTuple1, DottyTuple2...). */
+    object TupleApplies {
+      def unapply(tree: Apply)(implicit ctx: Context): Option[(List[Tree], TupleType)] =
+        tree match {
+          case Apply(TypeApply(Select(ident, nme.apply), fstTpe :: _), head :: tail :: Nil)
+            if ident.symbol == defn.TupleConsModule =>
+              tail match {
```

Comment:
It's probably better to introduce `TupleCons_apply` as a symbol in `Definitions`. Then you can test for it directly here:

    case Apply(TypeApply(fn, fstTpe :: _), head :: tail :: Nil) if fn.symbol == defn.TupleCons_apply